### DISCUSSION TITLE
Is there a way to construct a schema that matches the type of a template literal?

### DISCUSSION THREAD

**Discussion by willd-mwp** [colinhacks/zod#979] [Q&A] [answered]
Hi.

I've got a TS type that defines a structure of my form state, and creating a zod schema `z.ZodSchema<FormState>` to create a strictly typed validation schema for it.
On of the sub-properties of `FormState` looks something like this:
```typescript
logging: {
    loggingInterval: DurationISO,
    samplingInterval: DurationISO
}
```

where the `DurationISO` type is defined using a template literal that contains a union of string literals:
```typescript
type DurationISO = `PT${number}${"S" | "M" | "H"}`;
```
which can be expanded to be represented in this form:
```typescript
`PT${number}S` | `PT${number}M` | `PT${number}H`
```

In the validation schema for a property of this type, using `z.string()` throws a type error from the incompatibility of `string` vs ``PT${number}S` | `PT${number}M` | `PT${number}H``.
In `yup`, the solution to this is to provide a type argument to `yup.string()` like this `yup.string<DurationISO>()` and the type conflict gets resolved. I presume that Yup is try using this type argument to cast the string as the provided type rather than it altering the validation logic at all.
`zod` does not let you provide a type argument to `z.string()` though, so I was wondering if there was some other mechanism to work with types of this nature? For now i've just changed those sub-properties in `FormState` from `DurationISO` to `string`, but it's not an ideal solution.

**Comment by JacobWeisenburger** [comment, marked as answer]
You could do something like this.

Let me know if you have any questions.

```ts
type DurationISO = `PT${ number }${ 'S' | 'M' | 'H' }`
const durationISOSchema = z.string().regex( /^PT\d+(S|M|H)$/ ) as z.ZodSchema<DurationISO>

console.log( durationISOSchema.safeParse( 'PT1S' ).success ) // true
console.log( durationISOSchema.safeParse( 'PT12M' ).success ) // true
console.log( durationISOSchema.safeParse( 'PT123H' ).success ) // true

console.log( durationISOSchema.safeParse( 'PTH' ).success ) // false
console.log( durationISOSchema.safeParse( 'PT1.23' ).success ) // false
console.log( durationISOSchema.safeParse( 'PT12' ).success ) // false
console.log( durationISOSchema.safeParse( 'PT123' ).success ) // false
console.log( durationISOSchema.safeParse( 'P1' ).success ) // false
console.log( durationISOSchema.safeParse( 'foo' ).success ) // false
```

```ts
type Logging = {
    loggingInterval: DurationISO,
    samplingInterval: DurationISO
}

const loggingSchema: z.ZodSchema<Logging> = z.object( {
    loggingInterval: durationISOSchema,
    samplingInterval: durationISOSchema,
} )
```

**Reply by willd-mwp** [reply]
@JacobWeisenburger 

The ` as ZodSchema<DurationISO>` bit is the only bit I needed to solve my problem. I don't actually need to validate that the string matches the pattern at runtime, I just needed to avoid the type conflict at design-time. It completely slipped my mind that I could just cast the `string()` schema to my desired type by using the type argument of `ZodSchema<>`	.
Thanks
